### DISCUSSION TITLE
New lock creation in gen_rlock()

### DISCUSSION THREAD

**Discussion by mirza-ali-ctct** [elarivie/pyReaderWriterLock#14] [Q&A]
In the give example:
```python
with a.gen_rlock():
  #Read stuff

with a.gen_wlock():
  #Write stuff
```
do these `gen` functions create new `Lock` instance? I have a scenario where I wanna take read lock multiple times and wondering the implication of this lock creation approach

**Comment by elarivie** [comment]
Hello @mirza-ali-ctct 

To answer your question, `yes gen_* functions create a new lock instance`


a.gen_rlock() .. means generate a reader lock from the parent context 'a'
a.gen_wlock() .. means generate a writer lock from the parent context 'a'


For example the following creates two child reader lock and two child writer lock from the same parent context a:
```python
from readerwriterlock import rwlock
a = rwlock.RWLockFairD()  # This creates a context 
areader1 = a.gen_rlock()
areader2 = a.gen_rlock()
awriter1 = a.gen_wlock()
awriter2 = a.gen_wlock()
```

They are all related to the same context 'a'
They can all individually be lock/unlocked

Therefore:
```python
with awriter1:
    with awriter2:  # awriter2 won't ever be able to acquire a lock since it is already locked by awriter1
        # won't ever reach this
```

And Therefore:
```python
with areader1:
    with areader2:
        # This is reacheable because readerwriter locks allows more than one reader to lock at the same time
```

And Therefore:
```python
with areader1:
    with awriter1:  # awriter1 won't ever be able to acquire a lock since it is already locked by areader1
        # won't ever reach this
```
You could also create another context:
```
b = rwlock.RWLockFairD()
```
And the state of the locks from context 'a' and those from 'b' would not affect each other.

Hope it explains better the usage
